### PR TITLE
Support Computed Properties Capturing Dataset

### PR DESCRIPTION
**Pull request by flexyford**
In very rare cases in which the yielded dataset is captured as a computed property, we require a `forEach` array iteration implementation. An Ember Handlebars {{each}} helper can now be used by implementing `forEach` on the Collection Interface

Viewport implementations, such as smoke-and-mirrors` which capture the dataset as a computed property can now iterate over the dataset.

https://github.com/runspired/smoke-and-mirrors/blob/76594a08f91271c112cb7b9917576b74949567e6/addon/mixins/occlusion-collection.js#L913

Also snuck in some documentation improvements explaining the hijacked array iteration functions.

### CODE REVIEW

**Comment by cowboyd** [comment]
Test?

**Comment by flexyford** [comment]
I am not able to add a test using `smoke-and-mirrors` because it fails on teardown in the "after each" hook

File: addon/components/impagination-dataset.js [hunk 1]
```diff
@@ -98,38 +97,82 @@ var CollectionInterface = Ember.Object.extend(Ember.Array, {
...
     }
 
-    let length = end - start;
+    let length = end - begin;
 
     if (length < 0) {
       return [];
     }
-    Ember.run.schedule('afterRender', this, 'objectReadAt', start);
+    Ember.run.schedule('afterRender', this, 'objectReadAt', begin);
```
Comment (fivetanley) [inline addon/components/impagination-dataset.js]: Why does this need to know about afterRender?

File: addon/components/impagination-dataset.js [hunk 2]
```diff
@@ -98,38 +97,82 @@ var CollectionInterface = Ember.Object.extend(Ember.Array, {
...
+    return record;
+  },
+
+  // slice([begin[, end]])
+  // Advance the readOffset to `begin`
+  // | ========================================================================
+  // | Component / Helper | Release | URL                                      |
+  // | ------------------ |:------- | :--------------------------------------- |
+  // | Virtual-Each       | 0.2.0   | https://github.com/jasonmit/virtual-each |
+  // | ========================================================================
```
Comment (Robdel12) [inline addon/components/impagination-dataset.js]: 👏 great work here @flexyford

**Comment by flexyford** [comment]
Closing in Favor of #130

**Pull request closed without merging**